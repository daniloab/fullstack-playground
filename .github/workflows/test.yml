name: Tests

on:
  push:
    branches:
      - master
  pull_request:
    branches:
      - master

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2
      - name: Setup Node.js
        uses: actions/setup-node@v2
        with:
          node-version: '14'
      - run: yarn
      - run: yarn jest