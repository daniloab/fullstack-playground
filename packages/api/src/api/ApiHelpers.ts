import { Model, Types } from 'mongoose';
import { Context } from 'koa';

export const MESSAGE = {
  AUTH: {
    UNAUTHORIZED: 'Unauthorized',
  },
  COMMON: {
    INVALID_ID: 'Invalid id',
    EMAIL: 'email is a required field',
    EMAIL_INVALID: 'email must be a valid email',
    YUP_VALIDATION: 'Unknown error processing object',
  },
  LOGIN: {
    EMAIL_SUCCESS: 'Email successful',
    INCORRECT: 'Email or password incorrect',
    INVALID_LOGIN: 'Invalid login',
    INVALID_TENANT: 'Invalid tenant',
    SUCCESS: 'Login successful',
  },
  PAGE_INFO: {
    ERRORS: {
      NEGATIVE: 'Pagination values should be positive values',
    },
  },
  TENANT: {
    NOT_FOUND: 'Tenant not found',
  },
  USER: {
    CREATING: 'Some error occurred while creating user',
    MISSING: 'Missing user',
    NOT_FOUND: 'User not found',
    PASSWORD: 'You must informa a password',
    PRIMARY_KEY: 'User should have one of primary keys: id, or email',
    UPDATING: 'Some error occurred while updating user',
  },
};

export const getSkipAndLimit = (ctx: Context) => {
  const { skip = 0, limit = 100 } = ctx.query;

  if (skip < 0 || limit < 0) {
    return {
      skip: null,
      limit: null,
      errors: [{ data: { skip, limit }, message: MESSAGE.PAGE_INFO.ERRORS.NEGATIVE }],
    };
  }

  const mongoLimit = Math.min(parseInt(limit, 10), 100);
  const mongoSkip = parseInt(skip, 10);

  return {
    skip: mongoSkip,
    limit: mongoLimit,
    errors: null,
  };
};

type ErrorValidate = {
  data: {};
  message: string;
};

type PageInfo = {
  errors?: ErrorValidate[];
  skip: number;
  limit: number;
  totalCount: number;
  hasPreviousPage: number;
  hasNextPage: number;
};

export const getPageInfo = async (ctx: Context, model: Model): PageInfo => {
  const { company } = ctx;
  const { skip, limit, errors } = getSkipAndLimit(ctx);

  if (errors) {
    return {
      errors,
      skip,
      limit,
      totalCount: null,
      hasPreviousPage: null,
      hasNextPage: null,
    };
  }

  const conditionsTotalCount = {
    company,
    removedAt: null,
  };

  const totalCount = await model.count(conditionsTotalCount);

  const hasPreviousPage = skip > 0;
  const hasNextPage = skip + limit < totalCount;

  return {
    skip,
    limit,
    totalCount,
    hasPreviousPage,
    hasNextPage,
  };
};

export const checkObjectId = id => {
  if (!Types.ObjectId.isValid(id)) {
    return {
      error: true,
    };
  }

  return {
    error: false,
    _id: id,
  };
};
