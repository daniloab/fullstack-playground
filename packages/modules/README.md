# @fp/modules

All fullstack playground models/modules to be consumed by Api, GraphQL and so on 
