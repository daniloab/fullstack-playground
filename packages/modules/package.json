{
  "name": "@fullstack-playground/modules",
  "version": "1.0.0",
  "main": "index.js",
  "author": "Danilo Assis",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "src/index.ts",
  "devDependencies": {
    "@fullstack-playground/babel": "*",
    "@types/babel__core": "7.1.12"
  },
  "scripts": {
    "build": "babel src --extensions \".es6,.js,.es,.jsx,.mjs,.ts,.tsx\" --ignore *.spec.js --out-dir dist --copy-files --source-maps --verbose"
  },
  "dependencies": {
    "bcryptjs": "^2.4.3"
  }
}
