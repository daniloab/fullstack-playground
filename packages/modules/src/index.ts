export { default as TenantModel, ITenant } from './tenant/TenantModel';
export { createTenant } from './tenant/__fixtures__/createTenant';

export { default as UserModel, IUser } from './user/UserModel';
export { createUser } from './user/__fixtures__/createUser';
