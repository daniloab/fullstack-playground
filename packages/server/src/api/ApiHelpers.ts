import { Model } from 'mongoose';
import { Context } from 'koa';

export const MESSAGE = {
  AUTH: {
    UNAUTHORIZED: 'Unauthorized',
  },
  LOGIN: {
    EMAIL_SUCCESS: 'Email successful',
    INCORRECT: 'Email or password incorrect',
    INVALID_LOGIN: 'Invalid login',
    INVALID_TENANT: 'Invalid tenant',
    SUCCESS: 'Login successful',
  },
  PAGE_INFO: {
    ERRORS: {
      NEGATIVE: 'Pagination values should be positive values',
    },
  },
  TENANT: {
    NOT_FOUND: 'Tenant not found',
  },
  USER: {
    NOT_FOUND: 'User not found',
  },
};

export const getSkipAndLimit = (ctx: Context) => {
  const { skip = 0, limit = 100 } = ctx.query;

  if (skip < 0 || limit < 0) {
    return {
      skip: null,
      limit: null,
      errors: [{ data: { skip, limit }, message: MESSAGE.PAGE_INFO.ERRORS.NEGATIVE }],
    };
  }

  const mongoLimit = Math.min(parseInt(limit, 10), 100);
  const mongoSkip = parseInt(skip, 10);

  return {
    skip: mongoSkip,
    limit: mongoLimit,
    errors: null,
  };
};

type ErrorValidate = {
  data: {};
  message: string;
};

type PageInfo = {
  errors?: ErrorValidate[];
  skip: number;
  limit: number;
  totalCount: number;
  hasPreviousPage: number;
  hasNextPage: number;
};

export const getPageInfo = async (ctx: Context, model: Model): PageInfo => {
  const { company } = ctx;
  const { skip, limit, errors } = getSkipAndLimit(ctx);

  if (errors) {
    return {
      errors,
      skip,
      limit,
      totalCount: null,
      hasPreviousPage: null,
      hasNextPage: null,
    };
  }

  const conditionsTotalCount = {
    company,
    removedAt: null,
  };

  const totalCount = await model.count(conditionsTotalCount);

  const hasPreviousPage = skip > 0;
  const hasNextPage = skip + limit < totalCount;

  return {
    skip,
    limit,
    totalCount,
    hasPreviousPage,
    hasNextPage,
  };
};
