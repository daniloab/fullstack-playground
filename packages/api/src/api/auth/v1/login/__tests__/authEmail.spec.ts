import { clearDbAndRestartCounters, connectMongoose, disconnectMongoose } from '@fullstack-playground/test';
import { createTenant, createUser } from '@fullstack-playground/modules';

import { createApiCall } from '../../../../../../test';
import { MESSAGE } from '../../../../ApiHelpers';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

const url = '/api/auth/v1/login/email';

it('should return error if domain not found', async () => {
  const payload = {
    email: 'test@test.com',
  };

  const response = await createApiCall({ url, payload, domainname: 'test' });

  expect(response.body).toMatchSnapshot();
  expect(response.body.message).toBe(MESSAGE.TENANT.NOT_FOUND);
  expect(response.status).toBe(401);
});

it('should return error if tenant is inactivated', async () => {
  const payload = {
    email: 'test@test.com',
  };

  const response = await createApiCall({ url, payload, domainname: 'test' });

  expect(response.body).toMatchSnapshot();
  expect(response.body.message).toBe(MESSAGE.TENANT.NOT_FOUND);
  expect(response.status).toBe(401);
});

it('should return error if user not found', async () => {
  const tenant = await createTenant();
  const payload = {
    email: 'test@test.com',
  };

  const response = await createApiCall({ url, payload, domainname: tenant.domainName });

  expect(response.body).toMatchSnapshot();
  expect(response.body.message).toBe(MESSAGE.LOGIN.INCORRECT);
  expect(response.status).toBe(400);
});

it('should return success if user found', async () => {
  const tenant = await createTenant();
  const user = await createUser({ email: 'user@test.com', tenant });

  const payload = {
    email: user.email,
  };

  const response = await createApiCall({ url, payload, domainname: tenant.domainName });

  const { companyId, ...restBody } = response.body;

  expect(restBody).toMatchSnapshot();
  expect(response.body.message).toBe(MESSAGE.LOGIN.EMAIL_SUCCESS);
  expect(response.status).toBe(200);
});
