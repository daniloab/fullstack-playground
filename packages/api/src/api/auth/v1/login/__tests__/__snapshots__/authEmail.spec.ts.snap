// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return error if domain not found 1`] = `
Object {
  "message": "Tenant not found",
  "status": "ERROR",
}
`;

exports[`should return error if tenant is inactivated 1`] = `
Object {
  "message": "Tenant not found",
  "status": "ERROR",
}
`;

exports[`should return error if user not found 1`] = `
Object {
  "message": "Email or password incorrect",
  "status": "ERROR",
}
`;

exports[`should return success if user found 1`] = `
Object {
  "message": "Email successful",
}
`;
